Source/JavaScriptCore/runtime/JSCJSValue.h

```diff
@@ -86,8 +86,87 @@ enum PreferredPrimitiveType : uint8_t { NoPreference, PreferNumber, PreferString
 
 struct CallData;
 
+/*
+ * EncodedJSValueStorage is used as a storage type for JSValues that are
+ * accessible from multiple threads
+ *
+ * On 64-bit ports, and when JIT support is not being built,
+ * EncodedJSValueStorage is exactly int64_t.
+ *
+ * However, on 32-bit ports with JIT support, assigning to an
+ * EncodedJSValueStorage performs three writes: the first reserves the JSValue
+ * for writing, then the payload and tag are written (in that order)--this
+ * protocol enables JIT (and one day, GC mark) threads to perform an atomic read
+ * of the whole JSValue and retry if they observe the reserved state.
+ */
+#if USE(JSVALUE32_64) && ENABLE(JIT)
+struct alignas(8) EncodedJSValueStorage {
```

Comment:
Why not attaching `alignas(8)` to EncodedJSValue in 32bit environment? This makes code much cleaner.

Reply:
Yeah, it's unfortunate how much extra noise there is because this type isn't POD anymore.

Alignment alone isn't enough, though--I think we need runtime code that is touching JSValues in the heap to do the extra write of InvalidTag (and to perform the writes on `volatile` int32_ts); As far as I could tell, using this type as storage in WriteBarrier and friends achieves this nicely; I suppose an alternative approach would be to move this code to those places?

Reply:
> I suppose an alternative approach would be to move this code to those places?

I think that is better. We should minimize the use of`USE(JSVALUE32_64)` ifdefs as much as possible.